File: LeetCode/simplify_path.rb
# Given an absolute path for a file (Unix-style), simplify it.
#
# For example,
# path = "/home/", => "/home"
# path = "/a/./b/../../c/", => "/c"
#

def simplify_path(path)
    path_tokens = path.split("/").reject{ |x| x.empty? || x == "." }
    stack = []
    path_tokens.each{ |x|
        if x == ".."
            stack.pop if stack.any?
        else
            stack.push(x)
        end
    }

    "/" << stack.join("/")
end

["/home/", "/a/./b/../../c/", "/home//foo/"].each { |path| puts simplify_path(path) }

File: LeetCode/simplify_path.cs
using System;
using System.Collections.Generic;
using System.Linq;

static class Program {
    static String SimplifyPath(this String path) {
        Stack<String> stack = new Stack<String>();
        foreach (String element in path.Split(new char[] {'/'}, StringSplitOptions.RemoveEmptyEntries)) {
            switch (element) {
                case "..": if (stack.Count != 0) {
                               stack.Pop();
                           } 
                           
                           break;
                case ".": break;
                default: stack.Push(element);
                         break;
            }
        }

        return "/" + String.Join("/", stack.Reverse());
    }

    static void Main() {
        Console.WriteLine(String.Join(", ", new String[] {"/home/", "/a/./b/../../c/"}.Select(x => String.Format(" {0}={1} ", x, x.SimplifyPath()))));
    }
}
